### PR TITLE
Allow configuring the exemplar filter on the metrics SDK

### PR DESCRIPTION
**Pull request by dashpole**
Part of https://github.com/open-telemetry/opentelemetry-go/issues/5249

### Spec

https://opentelemetry.io/docs/specs/otel/metrics/sdk/#exemplarfilter

> The ExemplarFilter configuration MUST allow users to select between one of the built-in ExemplarFilters. While ExemplarFilter determines which measurements are eligible for becoming an Exemplar, the ExemplarReservoir makes the final decision if a measurement becomes an exemplar and is stored.

> The ExemplarFilter SHOULD be a configuration parameter of a MeterProvider for an SDK. The default value SHOULD be TraceBased. The filter configuration SHOULD follow the [environment variable specification](https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#exemplar).

> An OpenTelemetry SDK MUST support the following filters:

> * [AlwaysOn](https://opentelemetry.io/docs/specs/otel/metrics/sdk/#alwayson)
> * [AlwaysOff](https://opentelemetry.io/docs/specs/otel/metrics/sdk/#alwaysoff)
> * [TraceBased](https://opentelemetry.io/docs/specs/otel/metrics/sdk/#tracebased)

### Changes

* adds exemplar.AlwaysOffFilter, which is one of the required filters from the SDK: https://opentelemetry.io/docs/specs/otel/metrics/sdk/#alwaysoff
* adds `metric.WithExemplarFilter` as an option for the metrics SDK.
* moves handling of `OTEL_METRICS_EXEMPLAR_FILTER` to the same location as config handling to make code easier to navigate.

dropReservoir can actually be removed, but I plan to do that in a follow-up refactor, since it will be a large diff.

### CODE REVIEW

File: sdk/metric/config.go
Comment (dmathieu) [inline sdk/metric/config.go:155]:
```suggestion
// The exemplar filter determines which measurements are offered to the
```

**Review by dmathieu — APPROVED** [review]

File: sdk/metric/config.go
Comment (dashpole) [inline sdk/metric/config.go:180]:
```suggestion
		opts = append(opts, WithExemplarFilter(exemplar.TraceBasedFilter))
```

File: sdk/metric/config.go
Comment (dashpole) [inline sdk/metric/config.go:84]:
```suggestion
		exemplarFilter: exemplar.TraceBasedFilter,
```

File: sdk/metric/config.go
Comment (MrAlias) [inline sdk/metric/config.go:161]: nit: given `go.opentelemetry.io/otel/sdk/metric/exemplar` is imported already:

```suggestion
// By default, the [exemplar.SampledFilter]
// is used. Exemplars can be entirely disabled by providing the
// [exemplar.AlwaysOffFilter].
```

File: sdk/metric/config.go
Comment (MrAlias) [inline sdk/metric/config.go:89]: These are applied prior to options passed directly for the trace SDK:

https://github.com/open-telemetry/opentelemetry-go/blob/d37e1ef81c2f4482c7fb2ae686f3be7ffd4bca10/sdk/trace/provider.go#L99-L103

We should unify on a precedence.

File: sdk/metric/config.go
Comment (MrAlias) [inline sdk/metric/config.go:174]: Casing shouldn't matter here, right?

File: sdk/metric/config_test.go
Comment (MrAlias) [inline sdk/metric/config_test.go:197]: Missing test for the environment varaiable support.

File: sdk/metric/config.go
Comment (MrAlias) [reply sdk/metric/config.go:89]: SIG meeting note:

- Unify on the existing precedence: passed options take precedence over env var.

File: sdk/metric/config_test.go
Comment (dashpole) [reply sdk/metric/config_test.go]: Added: https://github.com/open-telemetry/opentelemetry-go/pull/5850/commits/020cf0fbe0c03e320e7cc4df0b4a406df2a0c5cf

File: sdk/metric/config.go
Comment (dashpole) [reply sdk/metric/config.go]: Made this match the trace sdk behavior: https://github.com/open-telemetry/opentelemetry-go/pull/5850/commits/92f4cd46f36e1f5159736d6d9af92124647d1230

**Review by MrAlias — APPROVED** [review]

**Comment by dashpole** [comment]
@MrAlias lets wait until after the release to merge.  I don't like merging public API changes right before a release.

**Review by hanyuancheung — APPROVED** [review]

**Pull request merged**
